californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -139,39 +159,38 @@ public BlockwiseLayer(final NetworkConfig config) {
 	@Override
 	public void sendRequest(final Exchange exchange, final Request request) {
 
-		BlockOption block2 = request.getOptions().getBlock2();
-		if (block2 != null && block2.getNum() > 0) {
-			// This is the case if the user has explicitly added a block option
-			// for random access.
-			// Note: We do not regard it as random access when the block num is
-			// 0. This is because the user might just want to do early block
-			// size negotiation but actually wants to receive all blocks.
-			LOGGER.fine("request contains block2 option, creating random-access blockwise status");
-			BlockwiseStatus status = new BlockwiseStatus(getSizeForSzx(block2.getSzx()), request.getOptions().getContentFormat());
-			status.setCurrentSzx(block2.getSzx());
-			status.setCurrentNum(block2.getNum());
-			status.setRandomAccess(true);
-			exchange.setResponseBlockStatus(status);
-			lower().sendRequest(exchange, request);
-
-		} else if (requiresBlockwise(request)) {
-			// This must be a large POST or PUT request
-			startBlockwiseUpload(exchange, request);
+		Request requestToSend = request;
 
-		} else {
-			// no blockwise transfer required
-			exchange.setCurrentRequest(request);
-			lower().sendRequest(exchange, request);
+		if (isTransparentBlockwiseHandlingEnabled()) {
+
+			BlockOption block2 = request.getOptions().getBlock2();
+			if (block2 != null && block2.getNum() > 0) {
+				// This is the case if the user has explicitly added a block option
+				// for random access.
+				// Note: We do not regard it as random access when the block number is 0.
+				// This is because the user might just want to do early block
+				// size negotiation but actually want to retrieve the whole body by means of
+				// a transparent blockwise transfer.
+				LOGGER.fine("outbound request contains block2 option, creating random-access blockwise status");
+				addRandomAccessBlock2Status(exchange, request, block2);
+
+			} else if (requiresBlockwise(request)) {
+				// This must be a large POST or PUT request
+				requestToSend = startBlockwiseUpload(exchange, request);
+
+			}
 		}
+
+		exchange.setCurrentRequest(requestToSend);
+		lower().sendRequest(exchange, requestToSend);
 	}
 
-	private void startBlockwiseUpload(final Exchange exchange, final Request request) {
+	private Request startBlockwiseUpload(final Exchange exchange, final Request request) {
 
-		BlockwiseStatus status = findRequestBlockStatus(exchange, request);
+		final KeyUri key = getKey(exchange, request);
+		Block1BlockwiseStatus status = getOutboundBlock1Status(key, exchange, request);
```

Comment:
> Probably not, but this is because you consider that there are 2 different "entities" (you call it client app) using the Californium client API and sharing the same CoAP server resource without knowing each other.

Ok, now we're getting somewhere. So you are arguing that the Californium API is NOT (and should not be) intended to be used by multiple *independent* clients, right? We might have had this constraint already in Californium (even if not by intention but by failure of implementation) so I cannot really argue against it. However, this also seems to be a problem of the fact that we are doing *transparent blockwise handling* within the stack instead of in the client (where it would belong IMHO). So, we could say that this case of concurrent block1 uploads needs to be handled by the client(s) until we have refactored the BlockwiseLayer out of the stack and into a support library to be employed by `CoapClient` and `CoapResource`. Until then we could follow your approach. WDYT?